File: schClass/SchData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Data;
using System.Data.SqlClient;
using System.Threading;
using System.Runtime.Serialization;

[assembly: log4net.Config.XmlConfigurator(ConfigFile = "log4net.config")]

namespace Algorithm
{

    //排程公用数据,每次排产前初始化数据
    [Serializable]
    public class SchData: ISerializable
    {
        public SchData()
        { 
        
        }
        //定义变量
        public string cVersionNo = "";
        public string cCalculateNo = "";  //排程运算号 用户名+ 时间

        public DateTime dtStart;        //     开始排程时间
        public DateTime dtEnd;          //     排程截止时间
        public DateTime dtToday;        //     当前时间，取数据库 


        public int iCurRows = 0;            // 排程当前任务数，用于统计当前进度 ,按工序行号为准
        public int iTotalRows = 100;        // 排程总任务数 
        public int iProgress = 0;           // 进度百分比
       // public ThreadTool iProgress = 0;           // 进度百分比
        

        //public static Logger logger = new Logger();           //日志记录对象       

        public DataTable dtSchProduct = new DataTable();           //排产产品
        public DataTable dtSchProductWorkItem = new DataTable();   //加工物料
        public DataTable dtResource = new DataTable();           //所有资源，有值    
        public DataTable dt_ResourceTime = new DataTable();      //资源正常工作日历
        public DataTable dt_ResourceSpecTime = new DataTable();      //资源特殊工作日历

        public string cSchCapType = "0";   // 排程产能方案  0 ---正常产能,1--加班产能,2--极限产能
   
        //public DataTable dtResResource = new DataTable();        //关键资源，按排产顺序排序 ，有值       
        public DataTable dtSchProductRoute = new DataTable();
        public DataTable dtSchProductRouteRes = new DataTable();
        public DataTable dtSchProductRouteItem = new DataTable();
        public DataTable dtSchProductRouteResTime = new DataTable();
        public DataTable dtSchResWorkTime = new DataTable();         //合并后资源日历时间段，排程完成后，写入t_SchResWorkTime

        public DataTable dtSchProductTemp = new DataTable();         //排产结果写回
        public DataTable dtSchProductRouteTemp = new DataTable();
        public DataTable dtSchProductRouteResTemp = new DataTable();

        public DataTable dtProChaType = new DataTable();        //工艺特征类型，有值
        public DataTable dtResChaValue = new DataTable();       //工艺特征值，有值
        public DataTable dtResChaCrossTime = new DataTable();   //工艺特征转换时间 ,有值
        
        //2016-10-21
        public DataTable dtWorkCenter = new DataTable();        //工作中心
        public DataTable dtDepartment = new DataTable();        //部门
        public DataTable dtTeam = new DataTable();              //班组
        public DataTable dtPerson = new DataTable();            //员工

        public DataTable dtItem = new DataTable();              // 物料档案  用于扩展
        public DataTable dtTechInfo = new DataTable();          // 工艺档案 用于扩展

        public DataTable dtTechLearnCurves = new DataTable();          //资源任务学习曲线       2017-11-23
        public DataTable dtResTechScheduSN = new DataTable();          //资源工艺特征排产顺序

        //所有资源列表 dtResource
        public List<Resource> ResourceList = new List<Resource>(10);

        //关键资源列表
        public List<Resource> KeyResourceList = new List<Resource>(10);

        //资源组列表
        public List<Resource> TeamResourceList = new List<Resource>(10);

        //再建立他们的联系
        //排程订单产品列表 dtSchProduct
        public List<SchProduct> SchProductList = new List<SchProduct>(10);

        //加工物料信息表
        public List<SchProductWorkItem> SchProductWorkItemList = new List<SchProductWorkItem>(10);
        

        //排程订单产品工艺模型 dtSchProductRoute
        public List<SchProductRoute> SchProductRouteList = new List<SchProductRoute>(10);

        //排程订单产品子料
        public List<SchProductRouteItem> SchProductRouteItemList = new List<SchProductRouteItem>(10);

        //工序对应的可用资源列表  dtSchProductRouteRes 
        public List<SchProductRouteRes> SchProductRouteResList = new List<SchProductRouteRes>(10);

        //任务占用资源时间段 t_SchProductRouteResTime
        public List<TaskTimeRange> TaskTimeRangeList = new List<TaskTimeRange>(10);

        //所有工作中心列表 dtWorkCenter
        public List<WorkCenter> WorkCenterList = new List<WorkCenter>(10);

        //所有加工物料档案列表 dtItem
        public List<Item> ItemList = new List<Item>(10);

        //所有工艺档案列表 dtTechInfo
        public List<TechInfo> TechInfoList = new List<TechInfo>(10);

        public string CVersionNo { get => cVersionNo; set => cVersionNo = value; }

       

        //排产算法扩展
        //[System.Xml.Serialization.XmlIgnore]
        ////[System.Web.Script.Serialization.ScriptIgnore]
        //[System.Runtime.Serialization.IgnoreDataMember]
        //public APSCustomer.AlgorithmExtend algorithmExtend ;


        //公用方法
        //1、两个日期对比，返回时间差,考虑资源工作日期，节假日跳过
        public static int GetDateDiff(Resource resource, string DatePart, DateTime dt1, DateTime dt2)
        { 
            Int32 liReturn = 0  ;

            if (resource == null || resource.cResourceNo == "") return 1;

                            
            DataRow[] drDaySelect ;
            drDaySelect = resource.schData.dt_ResourceTime.Select("cResourceNo = '" + resource.cResourceNo + "' and dPeriodBegTime >= '" + dt1.ToString() + "' and dPeriodBegTime <= '" + dt2.ToString() + "'");

            drDaySelect = drDaySelect.OrderBy(x => x["dPeriodDay"]).ToArray();

            DateTime dtDayTemp;


            //返回天
            if (DatePart == "d")
            {
                if (drDaySelect.Length > 0)
                {
                    dtDayTemp = (DateTime)drDaySelect[0]["dPeriodDay"];
                    liReturn = 1;

                    //foreach (DataRow item in drDaySelect)
                   
                    for (int i = 0; i < drDaySelect.Length; i++ )
                    {
                        if ((DateTime)drDaySelect[i]["dPeriodDay"] == dtDayTemp)
                            continue;
                        else
                        {
                            liReturn++;
                            dtDayTemp = (DateTime)drDaySelect[i]["dPeriodDay"];
                        }

                    }
                    
                }
            }
            

            return liReturn;
        }


        //公用方法
        //2、两个日期对比，返回时间差
        public static int GetDateDiff(string DatePart, DateTime dt1, DateTime dt2)
        {
            Int32 liReturn = 0;

            //被减数
            TimeSpan ts1 = new TimeSpan(dt1.Ticks);

            //减数
            TimeSpan ts2 = new TimeSpan(dt2.Ticks);

            TimeSpan ts = ts2.Subtract(ts1);//Duration()


            //返回秒
            if (DatePart == "s")
            {
                // liReturn = Int32.Parse(ts.TotalSeconds.ToString());
                liReturn = Convert.ToInt32((ts.TotalSeconds));
            } //返回小时
            //返回分钟
            if (DatePart == "m")
            {
                //liReturn = Int64.Parse(ts.TotalMinutes.ToString() );
                liReturn = Convert.ToInt32((ts.TotalMinutes));
            } //返回小时
            else if (DatePart == "h")
            {
                //liReturn = ts.Days * 24 + ts.Hours;
                //liReturn = Int32.Parse(ts.TotalHours.ToString());
                liReturn = Convert.ToInt32((ts.TotalHours));
            } //返回天
            else if (DatePart == "d")
            {
                //liReturn = Int32.Parse(ts.TotalDays.ToString());
                liReturn = Convert.ToInt32((ts.TotalDays));
            }


            return liReturn;
        }

        public static string GetDateDiffString(System.DateTime Date1, System.DateTime Date2, string Interval)
        {
            double dblYearLen = 365;//年的长度，365天   
            double dblMonthLen = (365 / 12);//每个月平均的天数   
            System.TimeSpan objT;
            objT = Date2.Subtract(Date1);


            switch (Interval)
            {
                case "y"://返回日期的年份间隔   
                    return System.Convert.ToInt32(objT.Days / dblYearLen).ToString();
                case "M"://返回日期的月份间隔   
                    return System.Convert.ToInt32(objT.Days / dblMonthLen).ToString();
                case "d"://返回日期的天数间隔   
                    return objT.TotalDays.ToString();
                case "h"://返回日期的小时间隔   
                    return objT.TotalHours.ToString();
                case "m"://返回日期的分钟间隔   
                    return objT.TotalMinutes.ToString();
                case "s"://返回日期的秒钟间隔   
                    return objT.TotalSeconds.ToString();
                case "ms"://返回时间的微秒间隔   
                    return objT.TotalMilliseconds.ToString();
                case "all":
                    return objT.ToString();
                default:
                    break;
            }
            return "0";
        }

        //3、增加日期
        public static DateTime AddDate(string DatePart, Int32 iAdd, DateTime dt1)
        {
            DateTime dtNew = dt1;

              //返回秒
            if (DatePart == "s")
            {
                dtNew = dt1.AddSeconds(iAdd);
            } //返回小时
            //返回分钟
            if (DatePart == "m")
            {
                dtNew = dt1.AddMinutes(iAdd);
            } //返回小时
            else if (DatePart == "h")
            {
                //liReturn = ts.Days * 24 + ts.Hours;
                dtNew = dt1.AddHours(iAdd);
            } //返回天
            else if (DatePart == "d")
            {
                dtNew = dt1.AddDays(iAdd);
            }
            else if (DatePart == "m")
            {
                dtNew = dt1.AddMonths(iAdd);
            }

            return dtNew;
        }

        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            //((ISerializable)SchProductList).GetObjectData(info, context);

            ((ISerializable)SchProductRouteList).GetObjectData(info, context);

            ((ISerializable)SchProductRouteResList).GetObjectData(info, context);

            ((ISerializable)TaskTimeRangeList).GetObjectData(info, context);
            
        }

    }
}

File: Algorithm/SchData.ts
import {
  Resource,
  TaskTimeRange,
  // IComparable,
  // ICloneable,
  ResSourceDayCap,
  SchProductRouteRes,
  SchParam,
  Comparer,
  DateTime,
  // TimeRangeAttribute,
  DataTable,
  DataRow,
  SchProduct,
  SchProductWorkItem,
  SchProductRoute,
  SchProductRouteItem,
  WorkCenter,
  Item,
  TechInfo,
} from './type'
export class SchData {
  cVersionNo: string = ''
  cCalculateNo: string = '' // 排程运算号 用户名+ 时间
  dtStart: DateTime // 开始排程时间
  dtEnd: DateTime // 排程截止时间
  dtToday: DateTime // 当前时间，取数据库
  iCurRows: number = 0 // 排程当前任务数，用于统计当前进度 ,按工序行号为准
  iTotalRows: number = 100 // 排程总任务数
  iProgress: number = 0 // 进度百分比
  cSchCapType: string = '0' // 排程产能方案  0 ---正常产能,1--加班产能,2--极限产能
  dtSchProduct: DataTable //= new DataTable(); //排产产品
  dtSchProductWorkItem: DataTable //= new DataTable(); //加工物料
  dtResource: DataTable //= new DataTable(); //所有资源，有值
  dt_ResourceTime: DataTable // = new DataTable(); //资源正常工作日历
  dt_ResourceSpecTime: DataTable // = new DataTable(); //资源特殊工作日历
  dtSchProductRoute: DataTable // = new DataTable();
  dtSchProductRouteRes: DataTable // = new DataTable();
  dtSchProductRouteItem: DataTable //= new DataTable();
  dtSchProductRouteResTime: DataTable //= new DataTable();
  dtSchResWorkTime: DataTable //= new DataTable(); //合并后资源日历时间段，排程完成后，写入t_SchResWorkTime
  dtSchProductTemp: DataTable //= new DataTable(); //排产结果写回
  dtSchProductRouteTemp: DataTable // = new DataTable();
  dtSchProductRouteResTemp: DataTable // = new DataTable();
  dtProChaType: DataTable //= new DataTable(); //工艺特征类型，有值
  dtResChaValue: DataTable //= new DataTable(); //工艺特征值，有值
  dtResChaCrossTime: DataTable // = new DataTable(); //工艺特征转换时间 ,有值
  dtWorkCenter: DataTable //= new DataTable(); //工作中心
  dtDepartment: DataTable //= new DataTable(); //部门
  dtTeam: DataTable //= new DataTable(); //班组
  dtPerson: DataTable //= new DataTable(); //员工
  dtItem: DataTable //= new DataTable(); // 物料档案  用于扩展
  dtTechInfo: DataTable //= new DataTable(); // 工艺档案 用于扩展
  dtTechLearnCurves: DataTable // = new DataTable(); //资源任务学习曲线       2017-11-23
  dtResTechScheduSN: DataTable //= new DataTable(); //资源工艺特征排产顺序
  ResourceList: Resource[] = new Array<Resource>(10)
  KeyResourceList: Resource[] = new Array<Resource>(10)
  TeamResourceList: Resource[] = new Array<Resource>(10)
  SchProductList: SchProduct[] = new Array<SchProduct>(10)
  SchProductWorkItemList: SchProductWorkItem[] = new Array<SchProductWorkItem>(
    10,
  )
  SchProductRouteList: SchProductRoute[] = new Array<SchProductRoute>(10)
  SchProductRouteItemList: SchProductRouteItem[] = new Array<
    SchProductRouteItem
  >(10)
  SchProductRouteResList: SchProductRouteRes[] = new Array<SchProductRouteRes>(
    10,
  )
  TaskTimeRangeList: TaskTimeRange[] = new Array<TaskTimeRange>(10)
  WorkCenterList: WorkCenter[] = new Array<WorkCenter>(10)
  ItemList: Item[] = new Array<Item>(10)
  TechInfoList: TechInfo[] = new Array<TechInfo>(10)
  constructor(config: any) {
    // this.dtSchProduct.datasource = config.datasource;
    // this.dtSchProductWorkItem.datasource = config.datasource;
    // this.dtResource.datasource = config.datasource;
    // this.dt_ResourceTime.datasource = config.datasource;
    // this.dtSchProductRoute.datasource = config.datasource;
    // this.dtSchProductRouteRes.datasource = config.datasource;
    // this.dtSchProductRouteResTime.datasource = config.datasource; //
    // this.dtProChaType.datasource = config.datasource;
    // this.dtSchProductRouteTemp.datasource = config.datasource;
    // this.dtResTechScheduSN.datasource = config.datasource;
    // this.dtTechInfo.datasource = config.datasource; //
  }
  static GetDateDiff(DatePart: string, dt1: DateTime, dt2: DateTime, aa: any): any {
    if (typeof DatePart == 'string') {
      let liReturn: number = 0
      let ts: number = dt2.getTime() - dt1.getTime()
      if (DatePart === 's') {
        liReturn = Math.floor(ts / 1000)
      } else if (DatePart === 'm') {
        liReturn = Math.floor(ts / (1000 * 60))
      } else if (DatePart === 'h') {
        liReturn = Math.floor(ts / (1000 * 60 * 60))
      } else if (DatePart === 'd') {
        liReturn = Math.floor(ts / (1000 * 60 * 60 * 24))
      }
      return liReturn
    } else if (typeof DatePart == 'object') {
      let resource: Resource = DatePart
      //@ts-ignore
      DatePart = dt1
      //@ts-ignore
      dt1 = dt2
      //@ts-ignore
      dt2 = aa
      //@ts-ignore
      let liReturn: number = 0
      if (!resource || resource.cResourceNo === '') return 1
      let drDaySelect: DataRow[] = resource.schData.dt_ResourceTime.Select(
        `cResourceNo = '${
          resource.cResourceNo
        }' and dPeriodBegTime >= '${dt1.toISOString()}' and dPeriodBegTime <= '${dt2.toISOString()}'`,
      )
      drDaySelect = drDaySelect.sort(
        (a, b) => a['dPeriodDay'] - b['dPeriodDay'],
      )
      let dtDayTemp: DateTime
      if (DatePart === 'd') {
        if (drDaySelect.length > 0) {
          dtDayTemp = drDaySelect[0]['dPeriodDay']
          liReturn = 1
          for (let i = 0; i < drDaySelect.length; i++) {
            if (drDaySelect[i]['dPeriodDay'] === dtDayTemp) continue
            else {
              liReturn++
              dtDayTemp = drDaySelect[i]['dPeriodDay']
            }
          }
        }
      }
      return liReturn
    }
  }

  static GetDateDiffString(Date1: DateTime, Date2: DateTime, Interval: string): string {
    const dblYearLen: number = 365 //年的长度，365天
    const dblMonthLen: number = 365 / 12 //每个月平均的天数
    const objT: number = Date2.getTime() - Date1.getTime()
    switch (Interval) {
      case 'y': //返回日期的年份间隔
        return Math.floor(objT / (dblYearLen * 24 * 60 * 60 * 1000)).toString()
      case 'M': //返回日期的月份间隔
        return Math.floor(objT / (dblMonthLen * 24 * 60 * 60 * 1000)).toString()
      case 'd': //返回日期的天数间隔
        return Math.floor(objT / (24 * 60 * 60 * 1000)).toString()
      case 'h': //返回日期的小时间隔
        return Math.floor(objT / (60 * 60 * 1000)).toString()
      case 'm': //返回日期的分钟间隔
        return Math.floor(objT / (60 * 1000)).toString()
      case 's': //返回日期的秒钟间隔
        return Math.floor(objT / 1000).toString()
      case 'ms': //返回时间的微秒间隔
        return objT.toString()
      case 'all':
        return objT.toString()
      default:
        break
    }
    return '0'
  }

  static AddDate(DatePart: string, iAdd: number, dt1: DateTime): DateTime {
    let dtNew: DateTime = new DateTime(dt1)
    if (DatePart === 's') {
      dtNew.setSeconds(dtNew.getSeconds() + iAdd)
    } else if (DatePart === 'm') {
      dtNew.setMinutes(dtNew.getMinutes() + iAdd)
    } else if (DatePart === 'h') {
      dtNew.setHours(dtNew.getHours() + iAdd)
    } else if (DatePart === 'd') {
      dtNew.setDate(dtNew.getDate() + iAdd)
    } else if (DatePart === 'M') {
      dtNew.setMonth(dtNew.getMonth() + iAdd)
    }
    return dtNew
  }

  public GetObjectData(info: any, context: any): void {
    ;(this.SchProductRouteList as any).GetObjectData(info, context)
    ;(this.SchProductRouteResList as any).GetObjectData(info, context)
    ;(this.TaskTimeRangeList as any).GetObjectData(info, context)
  }
  getDateDiff(r: any, d: any, beg: any, end: any): any {}
  GetDateDiff(...args) {
    // let a=[...args]
    //@ts-ignore
    return this.getDateDiff(...args)
  }
}
